### PR TITLE
[WIP] Implement test discovery on linux

### PR DESCRIPTION
**Pull request by aciidb0mb3r**
This is a WIP PR for implementing test discovery on linux using indexing
data. The basic idea is to leverage the IndexStore library to get the
test methods and then generate a LinuxMain.swift file during the build
process.

This PR "works" but there are a number of serious hacks that require
resolving before we can merge it.

### CODE REVIEW

File: Sources/Build/BuildDelegate.swift [hunk 1]
```diff
@@ -186,6 +186,104 @@ extension SPMLLBuild.Diagnostic: DiagnosticDataConvertible {
     }
 }
 
+class LinuxMainCommand: ExternalCommand {
```
Comment (jakepetroules) [inline Sources/Build/BuildDelegate.swift]: Please remove all occurrences of "Linux" from the names of these things -- there's nothing inherently Linux-specific about this so it should be named something like "TestMainCommand", "test-main-tool", etc.

File: Sources/SPMUtility/dlopen.swift [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+public enum DLError: Swift.Error {
+    case `open`(String)
+    case close(String)
+}
+
+public func dlopen(_ path: String?, mode: DLOpenFlags) throws -> DLHandle {
+  #if os(Windows)
+    guard let handle = path?.withCString(encodedAs: UTF16.self, LoadLibraryW) else {
```
Comment (jakepetroules) [inline Sources/SPMUtility/dlopen.swift]: *W 👍

File: Sources/SPMUtility/dlopen.swift [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  #endif
+    return unsafeBitCast(ptr, to: T.self)
+}
+
+public func dlclose(_ handle: DLHandle) throws {
+    try handle.close()
+}
+
+#if !os(Windows)
+public func dlerror() -> String? {
```
Comment (jakepetroules) [inline Sources/SPMUtility/dlopen.swift]: On Windows you can implement dlerror using GetLastError + FormatMessage. Example here: https://docs.microsoft.com/en-us/windows/desktop/debug/retrieving-the-last-error-code

File: Sources/SPMUtility/IndexStore.swift [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+                return classToMethods.map {
+                    TestCaseClass(name: $0.key, module: "", methods: $0.value.sorted())
+                }
+            }
+        }
+
+        let builder = Ref(TestCaseBuilder(), api: api)
+
+        let ctx = unsafeBitCast(Unmanaged.passUnretained(builder), to: UnsafeMutableRawPointer.self)
+        _ = fn.record_reader_occurrences_apply_f(recordReader, ctx) { ctx , occ -> Bool in
```
Comment (akyrtzi) [inline Sources/SPMUtility/IndexStore.swift]: What's the issue with using the functions that accept blocks ? \cc @benlangmuir

Comment (aciidb0mb3r) [reply Sources/SPMUtility/IndexStore.swift]: There's a bug in Swift compiler that leads to a crash on linux if the block return a bool :( <rdar://problem/51629934>

Comment (akyrtzi) [reply Sources/SPMUtility/IndexStore.swift]: Ah, that's unfortunate.

**Comment by aciidb0mb3r** [comment]
@swift-ci smoke test

File: Sources/Build/BuildDelegate.swift [hunk 5]
```diff
@@ -186,6 +186,183 @@ extension SPMLLBuild.Diagnostic: DiagnosticDataConvertible {
...
+    func getSignature(_ command: SPMLLBuild.Command) -> [UInt8] {
+        return []
+    }
+
+    func execute(_ command: SPMLLBuild.Command) -> Bool {
+        fatalError("subclass responsibility")
+    }
+}
+
+final class TestDiscoveryCommand: CustomLLBuildCommand {
```
Comment (ddunbar) [inline Sources/Build/BuildDelegate.swift]: Docs would be nice, I have no idea what this command is trying to do :)

Comment (aciidb0mb3r) [reply Sources/Build/BuildDelegate.swift]: Going to add some docs in a separate commit.

**Comment by weissi** [comment]
@aciidb0mb3r would that be something we could have on 5.1 too?

**Comment by aciidb0mb3r** [comment]
Ya, I think that would be fine since this is currently opt-in.

**Comment by weissi** [comment]
> Ya, I think that would be fine since this is currently opt-in.

Awesome. Are you doing the backport cherry-pick or do you want me to?

**Comment by aciidb0mb3r** [comment]
I am just bringing the 5.1 branch in sync with master as I plan on making more bug fixes (in unrelated areas) that might make cherry-picking harder: https://github.com/apple/swift-package-manager/pull/2196

File: Sources/SPMUtility/IndexStore.swift (hunk 4 above)
Comment (kastiglione) [reply Sources/SPMUtility/IndexStore.swift]: Is there a bugs.swift.org for this? For reference, I have a PR for compiling libIndexStore with `-fblocks` on linux, and I'm wondering if that PR should be blocked on this bug. In our testing on linux, I don't believe we saw any crashes. https://github.com/apple/swift-clang/pull/291

**Comment by weissi** [comment]
CC @tomerd  btw

**Comment by dabrahams** [comment]
Is there a way we can turn this on in a `Package.swift` so that nobody needs to add the flag explicitly?
@compnerd does this feature work on Windows now?

**Comment by benlangmuir** [comment]
@dabrahams there's a pitch to enable this by default, and the current thinking is to enable it automatically when there is no LinuxMain.swift file present. https://forums.swift.org/t/pitch-enable-test-discovery-by-default/36619/15

**Comment by dabrahams** [comment]
@benlangmuir So is there no way for me to turn it on in a `Package.swift` file?

**Comment by SDGGiesbrecht** [comment]
@dabrahams,

> does this feature work on Windows now?

SwiftPM does not exist in the Windows toolchain yet; very little works (mostly because of path handling). That said, there is nothing platform‐specific about this feature, so I imagine it will just work on Windows once the lower‐level issues are resolved. (It certainly worked for Android without anyone needing to put any effort to supporting it.)

> So is there no way for me to turn it on in a Package.swift file?

No.

**Comment by dabrahams** [comment]
Thanks all for the replies.  Turns out I can't use test discovery anyway because swiftpm is incompatible with my company's build system restrictions (can't download stuff).  Someday.

**Pull request merged**